### ISSUE TITLE
Cannot project fault solution field to derived field

### ISSUE THREAD

**Issue by baagaard-usgs** [geodynamics/pylith#646]
## Describe the bug

When we want to output the change in fault tractions for prescribed slip, we want to project the Lagrange multiplier field to a derived field (`traction_change`). This type of operation works with `DMProjectField()` which we use to output the Lagrange multiplier field directly but does not appear to work for `DMProjectFieldLocal()` which we use when we compute a derived field.
It looks like the wrong DS is used to get the solution field. We get 1 solution field, which appears to be the displacement. For a fault kernel, I thought we should be getting all solution fields (displacement and Lagrange multiplier).

## To Reproduce

**PETSc branch**: `knepley/pylith`
**PyLith fork**: `https://github.com/baagaard-usgs/pylith/`
**PyLith branch**: `feature-output-traction-change`

Steps to reproduce the behavior.

```
# Build and then in the build tree...
cd tests/fullscale/linearelasticity/faults-2d
pylith shearnoslip.cfg shearnoslip_tri.cfg
```

The kernel for computing the derived field is [tractionChange_asVector()](https://github.com/baagaard-usgs/pylith/blob/feature-output-traction-change/libsrc/pylith/fekernels/FaultCohesiveKin.hh#L587).

**Comment by baagaard-usgs** [comment]
@knepley I pushed this change to pass in the label to restrict the projection "input" (solution field in this case). Currently, I get zeros in the output; I think you said this would happen because the label is not in the projection "output".

```
--- a/libsrc/pylith/feassemble/IntegratorInterface.cc
+++ b/libsrc/pylith/feassemble/IntegratorInterface.cc
@@ -808,7 +808,12 @@ pylith::feassemble::IntegratorInterface::_computeDerivedField(const PylithReal t
     PetscInt labelValue = 0;
     const PetscInt part = 0;
     err = DMSetAuxiliaryVec(derivedDM, dmLabel, labelValue, part, _auxiliaryField->getLocalVector());PYLITH_CHECK_ERROR(err);
+#if 0
     err = DMProjectFieldLocal(derivedDM, t, solution.getLocalVector(), kernelsArray, INSERT_VALUES, _derivedField->getLocalVector());PYLITH_CHECK_ERROR(err);
+#else
+    err = DMGetLabel(solution.getDM(), _labelName.c_str(), &dmLabel);PYLITH_CHECK_ERROR(err);
+    err = DMProjectFieldLabelLocal(derivedDM, t, dmLabel, 1, &_labelValue, PETSC_DETERMINE, NULL, solution.getLocalVector(), kernelsArray, INSERT_VALUES, _derivedField->getLocalVector());PYLITH_CHECK_ERROR(err);
+#endif
```

**Comment by knepley** [comment]
@baagaard-usgs Hmm, this will not solve it because the label does not intersect with the output mesh at all.

The cohesive label marks cohesive cells and cohesive faces, but not the faces on the edges of the fault. This seems right, but then how do we connect faces of the fault with the cohesive DS? We could mark those edge faces, but that does not seem right. I think the problem is the construction of the fault mesh we are using. The cells hanging off are normal cells, but we really want cohesive cells to hang off. Where is this fault mesh made? We should change it to have cohesive cells.

**Comment by baagaard-usgs** [comment]
This basic process already works when we are just projecting the Lagrange multiplier field to the fault mesh. The difference in this case is that we have not extracted just the Lagrange multiplier field before projecting. In this case we also want all solution fields for the derived field kernel, like we do for the residual kernels. This will allow us to compute slip in the derived kernel.

**Comment by knepley** [comment]
Right. The Lagrange multiplier works because it has the same discretization as the displacement. If they were different, that would fail as well. I think the right thing is to have a fault mesh that incorporates the cells with the discretization we want. That makes the most sense to me. I will just remake that mesh choosing the other side.
